### PR TITLE
Rather than merge array contents, replace the array

### PR DESCRIPTION
**Pull request by jakehoward**
@cliftonc @BorePlusPlus 

@cliftonc - can you help me merge and promote this? todo:
- Bump the major version number, this changes the API
- Decide what to do about the etcd tests, I can't get them to run properly and neither can @BorePlusPlus 
- Push to npm (looks like we do it manually? Can't find a travis file or Jenkins job)

I then plan to bump the major version in electric-conflab and upgrade there.

Motivation: running database migrations in the test environment to clean up the database, this way you don't need to hand craft a connection before the app has started.

## NB
- I wasn't able to run the etcd tests, but I did pin the etcd docker image to a version that supports rmdir.

### CODE REVIEW

**Comment by cliftonc** [comment]
Why do we want to change this?  I'm trying to recall the reason it was done this way, but it was a much younger and more adventurous version of me who made that decision.

File: index.js [hunk 1]
```diff
@@ -12,7 +12,11 @@ var minimist = require('minimist');
 var pathval = require('pathval');
 var stripBom = require('strip-bom');
 var defaultsDeep = _.partialRight(_.merge, function deep(value, other) {
-  return _.merge(value, other, deep);
+    if (_.isPlainObject(value) && _.isPlainObject(other)) {
+        return _.merge(value, other, deep);
+    }
+
+    return value;
```
Comment (jakehoward) [inline index.js]: Would probably expect this to return "other" but the way it's used, the value we want is on the left (I think). The tests pass this way around... Am I losing the plot?

**Comment by jakehoward** [comment]
@cliftonc - rehashed it a bit, the defaultsDeep was too surprising for the next person.

To answer your question: I'm not sure why it is the way it is at the moment, probably just because it's the default of `_.merge`. If you have a default config that has say: `tasks: ['a', 'b', 'c']` and you have a live config of `tasks: ['c', 'd']` you're not going to want: `tasks: ['c', 'd', 'c']` to be your final config. If you're lucky, you have the same number of items in your array and it's all good. If not, you get a surprise (even more so if it's an array of objects).

If someone is using this, they're playing a dangerous game because it's dependent on the order of the array. I can image you could use it for rascal perhaps, but I suggest we bump the major version and update anything that breaks (if they update electric conflab major version).

**Comment by cliftonc** [comment]
Interesting, I would have assumed it concatenated them so you would get all the array entries.  I guess this fits the model that happens with simple values, so works for me.

**Comment by cliftonc** [comment]
Interesting, I would have assumed it concatenated them so you would get all the array entries.  I guess this fits the model that happens with simple values, so works for me.

**Comment by BorePlusPlus** [comment]
>Interesting, I would have assumed it concatenated them so you would get all the array entries.

By default it treats arrays same as objects with indexes as keys - hence the same merging semantics.

Side note: As there are many decisions one needs to take when merging objects, @cressie176 wrote a module (of course he did) that lets you specify rules for most (all?) of these things.

**Comment by cressie176** [comment]
What you really want is a [config manager that supports custom merge rules](https://github.com/guidesmiths/confabulous/commit/94eb28a164b89f73271c011c6d2ab7494de35ba5)

What was that disappear emoticon again?

**Pull request merged**